### PR TITLE
Refactor: Validation: Make parts of IsValidPeginWitness templated

### PR DESCRIPTION
**Pull request by jtimon**
This should not change behavior in any way.

Separated out of https://github.com/ElementsProject/elements/pull/362

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2359,6 +2359,12 @@ CScript calculate_contract(const CScript& federationRedeemScript, const CScript&
     return scriptDestination;
 }
 
+bool GetAmountFromParentChainPegin(CAmount& amount, const Sidechain::Bitcoin::CTransaction& txBTC, unsigned int nOut)
+{
+    amount = txBTC.vout[nOut].nValue;
+    return true;
```
[the remark this answers is not in the document]
Comment (jtimon) [reply src/validation.cpp]: The point is to make the interface like https://github.com/ElementsProject/elements/pull/362/commits/23af08444f59fa1ce6a4de0c1b16e43a74ebf98d#diff-24efdb00bfbe56b140fb006b562cc70bR2369 which can fail.

**Comment by instagibbs** [comment]
utACK https://github.com/ElementsProject/elements/pull/370/commits/20f9fb3e5254ffcaa8ddf84dba5e1a58529f9cda

I'd like another set of eyes on this as well.

**Comment by jtimon** [comment]
Fixed nits

File: src/validation.cpp [hunk 2]
```diff
@@ -2359,6 +2359,78 @@ CScript calculate_contract(const CScript& federationRedeemScript, const CScript&
...
+}
+
+template<typename T>
```
Comment (jhfrontz) [inline src/validation.cpp]: I'm curious about the use of a template here (vs., say, an inline static function).  This isn't a header file and there only appears to be one use of this template/function -- is this a project coding convention or is there some performance/maintenance gain?

**Comment by instagibbs** [comment]
re-utACK https://github.com/ElementsProject/elements/pull/370/commits/a20986b40dd8330510fc8970f48d8ad620141adb

File: src/validation.cpp (hunk 2 above)
Comment (mword) [reply src/validation.cpp]: I'm not the biggest fan of adding code for future use into a pull request, it could sit there making no sense for a long time.

**Comment by mword** [comment]
I built and ran all tests (including qa tests and the peggin.py test) with no errors. I made some comments in #362 that are relevant to #370 I think (362 seemed the better place to make my comments). I got to the point where I understood the code if not the higher level logic that makes these changes necessary (I'm still coming to terms with that).

All in all: tACK

**Comment by mword** [comment]
ACK

File: src/validation.cpp (hunk 2 above)
Comment (jtimon) [reply src/validation.cpp]: Yeah, I did it only because @instagibbs asked. Perhaps I should have done 2 commits but not 2 PRs

**Pull request merged**